packages/contracts/contracts/universal/TokenMultiSigWallet.sol

```diff
@@ -0,0 +1,238 @@
+        confirms.confirmationCount -= getVotes(msg.sender);
+        emit ConfirmationRevoked(msg.sender, _transactionId);
+    }
+
+    /**
+     * @inheritdoc ITokenMultiSigWallet
+     */
+    function executeTransaction(uint256 _transactionId)
+        public
+        nonReentrant
+        transactionExists(_transactionId)
+    {
+        require(isConfirmed(_transactionId), "TokenMultiSigWallet: quorum not reached");
+        require(!transactions[_transactionId].executed, "TokenMultiSigWallet: already executed");
+
+        Types.MultiSigTransaction storage txn = transactions[_transactionId];
+        txn.executed = true;
+        bool success = SafeCall.call(txn.target, gasleft(), txn.value, txn.data);
+        require(success, "TokenMultiSigWallet: call transaction failed");
+        emit TransactionExecuted(msg.sender, _transactionId);
+    }
+
+    /**
+     * @inheritdoc ITokenMultiSigWallet
+     */
+    function isConfirmed(uint256 _transactionId) public view returns (bool) {
+        return confirmations[_transactionId].confirmationCount >= quorum();
+    }
+
+    /**
+     * @inheritdoc ITokenMultiSigWallet
+     */
+    function quorum() public view returns (uint256) {
+        uint256 currentTimepoint = clock() - 1;
+        return
+            (IERC5805Upgradeable(address(GOVERNOR.token())).getPastTotalSupply(currentTimepoint) *
+                GOVERNOR.quorumNumerator(currentTimepoint)) / GOVERNOR.quorumDenominator();
+    }
+
+    /**
+     * @inheritdoc ITokenMultiSigWallet
+     */
+    function getVotes(address account) public view returns (uint256) {
+        return IERC5805Upgradeable(address(GOVERNOR.token())).getVotes(account);
+    }
+
+    /**
+     * @notice Returns whether the account has confirmed the transaction.
+     *
+     * @param _transactionId Index of transaction to check.
+     * @param _account       Address to check.
+     */
+    function isConfirmedBy(uint256 _transactionId, address _account) public view returns (bool) {
+        return confirmations[_transactionId].confirmedBy[_account];
+    }
+
+    function getConfirmationCount(uint256 _transactionId) public view returns (uint256) {
+        return confirmations[_transactionId].confirmationCount;
+    }
+
+    /**
+     * @notice Generate id of the transaction.
+     *
+     * @param _target Transaction target address.
+     * @param _value  Transaction ether value.
+     * @param _data   Transaction data payload.
+     */
+    function generateTransactionId(
+        address _target,
+        uint256 _value,
+        bytes memory _data
+    ) public view validAddress(_target) returns (uint256) {
+        return uint256(keccak256(abi.encode(_target, _value, _data, clock())));
+    }
+
+    /**
+     * @dev Clock (as specified in EIP-6372) is set to match the token's clock.
+     *      Fallback to block numbers if the token does not implement EIP-6372.
+     */
+    function clock() public view returns (uint48) {
```

Comment:
I wonder why these functions are not included in the interface.

Reply:
isConfirmedBy : added
getConfirmationCount : added
generateTransactionId : This is not a function used in the general case.
clock : This is not a function used in the general case.